spec/lib/theme_settings_object_validator_spec.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+RSpec.describe ThemeSettingsObjectValidator do
+  describe "#validate" do
+    it "should return the right array of error messages when properties are required but missing" do
+      schema = {
+        name: "section",
+        properties: {
+          title: {
+            type: "string",
+            required: true,
+          },
+          description: {
+            type: "string",
+            required: true,
+          },
+          links: {
+            type: "objects",
+            schema: {
+              name: "link",
+              properties: {
+                name: {
+                  type: "string",
+                  required: true,
+                },
+                child_links: {
+                  type: "objects",
+                  schema: {
+                    name: "child_link",
+                    properties: {
+                      title: {
+                        type: "string",
+                        required: true,
+                      },
+                      not_required: {
+                        type: "string",
+                      },
+                    },
+                  },
+                },
+              },
+            },
+          },
+        },
+      }
+
+      error_messages = described_class.new(schema:, object: {}).validate
+
+      expect(error_messages).to eq(
+        ["title property must be present", "description property must be present"],
+      )
+
+      error_messages =
+        described_class.new(
+          schema: schema,
+          object: {
+            links: [{ child_links: [{}, {}] }, {}],
+          },
+        ).validate
+
+      expect(error_messages).to eq(
+        [
+          "title property must be present",
+          "description property must be present",
+          "name property must be present in 'link' object 1",
+          "title property must be present in 'child_link' object 1",
+          "title property must be present in 'child_link' object 2",
+          "name property must be present in 'link' object 2",
+        ],
+      )
```

Comment:
@OsamaSayegh I'm not sure how far you are with the front end stuff but I'm wondering if this format is sufficient for you.